src/graph.cc

```diff
@@ -641,6 +655,68 @@ bool ImplicitDepLoader::LoadDepsFromLog(Edge* edge, string* err) {
   return true;
 }
 
+bool ImplicitDepLoader::LoadOutputsFromLog(Edge* edge, string* err) {
+  // NOTE: deps are only supported for single-target edges.
+  Node* output = edge->outputs_[0];
+  DepsLog::Deps* deps = deps_log_ ? deps_log_->GetDeps(output) : NULL;
+  if (!deps) {
+    EXPLAIN("outputs for '%s' are missing", output->path().c_str());
+    return false;
+  }
+
+  // Deps are invalid if the output is newer than the deps.
+  if (output->mtime() > deps->mtime) {
+    EXPLAIN("stored outputs info out of date for '%s' (%" PRId64 " vs %" PRId64
+            ")",
+            output->path().c_str(), deps->mtime, output->mtime());
+    return false;
+  }
+
+  vector<Node*> implicit_outputs;
+  implicit_outputs.reserve(deps->outputs_count);
+
+  for (size_t i = deps->node_count - deps->outputs_count; i < deps->node_count; i++) {
+    bool exist = false;
+    Node* new_node = deps->nodes[i];
+    for (vector<Node*>::iterator n = edge->outputs_.begin();
+         n != edge->outputs_.end(); n++) {
+      if (*n == new_node) {
+        exist = true;
+        break;
+      }
+    }
+    if (!exist) {
+      if (!new_node->StatIfNecessary(disk_interface_, err)) {
+        return false;
+      }
+      implicit_outputs.push_back(new_node);
+    }
+  }
+
+  // Add the dyndep-discovered outputs to the edge.
+  edge->outputs_.insert(edge->outputs_.end(), implicit_outputs.begin(),
+                        implicit_outputs.end());
+  edge->implicit_outs_ += implicit_outputs.size();
+
+  // Add this edge as incoming to each new output.
+  for (std::vector<Node*>::const_iterator i = implicit_outputs.begin();
+       i != implicit_outputs.end(); ++i) {
+    if (Edge* old_in_edge = (*i)->in_edge()) {
+      // This node already has an edge producing it.  Fail with an error
+      // unless the edge was generated by ImplicitDepLoader, in which
+      // case we can replace it with the now-known real producer.
+      if (!old_in_edge->generated_by_dep_loader_) {
+        *err = "multiple rules generate " + (*i)->path();
+        return false;
+      }
+      old_in_edge->outputs_.clear();
+    }
+    (*i)->set_in_edge(edge);
+  }
+
+  return true;
```

Comment:
Hi again, I'm aware that this PR is outdated, but that's what we based our version of Ninja on, so we have some interest to fix this one here. However, we do want to contribute back our insights that may help others and are of course interested in receiving feedback if we miss something.


```suggestion
  bool hasConflict = false;
  for (std::vector<Node*>::const_iterator i = implicit_outputs.begin();
       i != implicit_outputs.end(); ++i) {
    if (Edge* old_in_edge = (*i)->in_edge()) {
      // This node already has an edge producing it.  
      // This can mean that there is a conflict of two build statements producing the same output,
      // but it can also mean that the dynout information within .ninja_deps file is outdated and
      // the dynamic output is now an explicit output of another build statement.
      // We can not be sure here if the conflict is real or not, so we return false to mark the output as dirty.
      // The conflict detection needs to happen when processing the dynout file from within Builder::FinishCommand.
      EXPLAIN("Dynamic output '%s' of a previous execution can now be created via '%s', so we consider '%s' as dirty.", 
          (*i)->path().c_str(), old_in_edge->rule_->name().c_str(), output->path().c_str());
      hasConflict = true;
    } else {
      edge->outputs_.push_back(*i);
      edge->implicit_outs_++;
      (*i)->set_in_edge(edge);
    }
  }

  return !hasConflict;
```

Reply:
Coming back to this one. In the meantime we discovered this does not work either in all situations. For now, we discontinued working on this one.